### PR TITLE
wip – feat: viem

### PR DESCRIPTION
**Pull request by jxom**
[Migration Notes](https://github.com/wagmi-dev/wagmi/blob/jxom/viem/notes.md)

### Code

- [x] `@wagmi/core` 
  - [x] Actions
  - [x] Types
  - [x] Utilities
  - [x] Migrate `configureChains` 
  - [x] Remove redundant modules
  - [x] Rename "Provider" & "Signer" terminology to "Public Client" & "Wallet Client"
- [x] `wagmi` 
  - [x] Hooks
  - [x] Utilities
  - [x] Remove redundant modules
  - [x] Rename "Provider" & "Signer" terminology to "Public Client" & "Wallet Client"
- [x] [`@wagmi/connectors`](https://github.com/wagmi-dev/references/pull/215)
- [x] CLI (React)

### Docs

- [x] Update Examples
- [x] `core/chains.en-US.mdx`
- [x] `core/client.en-US.mdx`
- [x] `core/getting-started.en-US.mdx`
- [x] `core/actions/connect.en-US.mdx`
- [x] `core/actions/disconnect.en-US.mdx`
- [x] `core/actions/fetchBalance.en-US.mdx`
- [x] `core/actions/fetchBlockNumber.en-US.mdx`
- [x] `core/actions/fetchEnsAddress.en-US.mdx`
- [x] `core/actions/fetchEnsAvatar.en-US.mdx`
- [x] `core/actions/fetchEnsName.en-US.mdx`
- [x] `core/actions/fetchEnsResolver.en-US.mdx`
- [x] `core/actions/fetchFeeData.en-US.mdx`
- [x] `core/actions/fetchSigner.en-US.mdx`
- [x] `core/actions/fetchToken.en-US.mdx`
- [x] `core/actions/fetchTransaction.en-US.mdx`
- [x] `core/actions/getAccount.en-US.mdx`
- [x] `core/actions/getContract.en-US.mdx`
- [x] `core/actions/getNetwork.en-US.mdx`
- [x] `core/actions/getProvider.en-US.mdx`
- [x] `core/actions/getWebSocketProvider.en-US.mdx`
- [x] `core/actions/multicall.en-US.mdx`
- [x] `core/actions/prepareSendTransaction.en-US.mdx`
- [x] `core/actions/prepareWriteTransaction.en-US.mdx`
- [x] `core/actions/readContract.en-US.mdx`
- [x] `core/actions/readContracts.en-US.mdx`
- [x] `core/actions/sendTransaction.en-US.mdx`
- [x] `core/actions/signMessage.en-US.mdx`
- [x] `core/actions/signTypedData.en-US.mdx`
- [x] `core/actions/switchNetwork.en-US.mdx`
- [x] `core/actions/waitForTransaction.en-US.mdx`
- [x] `core/actions/watchAccount.en-US.mdx`
- [x] `core/actions/watchBlockNumber.en-US.mdx`
- [x] `core/actions/watchContractEvent.en-US.mdx`
- [x] `core/actions/watchMulticall.en-US.mdx`
- [x] `core/actions/watchNetwork.en-US.mdx`
- [x] `core/actions/watchPendingTransactions.en-US.mdx`
- [x] `core/actions/watchProvider.en-US.mdx`
- [x] `core/actions/watchReadContract.en-US.mdx`
- [x] `core/actions/watchReadContracts.en-US.mdx`
- [x] `core/actions/watchSigner.en-US.mdx`
- [x] `core/actions/watchWebSocketProvider.en-US.mdx`
- [x] `core/actions/writeContract.en-US.mdx`
- [x] `core/connectors/coinbaseWallet.en-US.mdx`
- [x] `core/connectors/injected.en-US.mdx`
- [x] `core/connectors/ledger.en-US.mdx`
- [x] `core/connectors/metaMask.en-US.mdx`
- [x] `core/connectors/mock.en-US.mdx`
- [x] `core/connectors/safe.en-US.mdx`
- [x] `core/connectors/walletConnect.en-US.mdx`
- [x] `core/connectors/walletConnectLegacy.en-US.mdx`
- [x] `core/providers/alchemy.en-US.mdx`
- [x] `core/providers/configuring-chains.en-US.mdx`
- [x] `core/providers/infura.en-US.mdx`
- [x] `core/providers/jsonRpc.en-US.mdx`
- [x] `core/providers/public.en-US.mdx`
- [x] `react/actions.en-US.mdx`
- [x] `react/chains.en-US.mdx`
- [x] `react/comparison.en-US.mdx`
- [x] `react/getting-started.en-US.mdx`
- [x] `react/prepare-hooks.en-US.mdx`
- [x] `react/WagmiConfig.en-US.mdx`
- [x] `react/connectors/coinbaseWallet.en-US.mdx`
- [x] `react/connectors/injected.en-US.mdx`
- [x] `react/connectors/ledger.en-US.mdx`
- [x] `react/connectors/metaMask.en-US.mdx`
- [x] `react/connectors/mock.en-US.mdx`
- [x] `react/connectors/safe.en-US.mdx`
- [x] `react/connectors/walletConnect.en-US.mdx`
- [x] `react/connectors/walletConnectLegacy.en-US.mdx`
- [x] `react/hooks/useAccount.en-US.mdx`
- [x] `react/hooks/useBalance.en-US.mdx`
- [x] `react/hooks/useBlockNumber.en-US.mdx`
- [x] `react/hooks/useConnect.en-US.mdx`
- [x] `react/hooks/useContract.en-US.mdx`
- [x] `react/hooks/useContractEvent.en-US.mdx`
- [x] `react/hooks/useContractInfiniteReads.en-US.mdx`
- [x] `react/hooks/useContractRead.en-US.mdx`
- [x] `react/hooks/useContractReads.en-US.mdx`
- [x] `react/hooks/useContractWrite.en-US.mdx`
- [x] `react/hooks/useDisconnect.en-US.mdx`
- [x] `react/hooks/useEnsAvatar.en-US.mdx`
- [x] `react/hooks/useEnsAddress.en-US.mdx`
- [x] `react/hooks/useEnsResolver.en-US.mdx`
- [x] `react/hooks/useFeeData.en-US.mdx`
- [x] `react/hooks/useNetwork.en-US.mdx`
- [x] `react/hooks/useProvider.en-US.mdx`
- [x] `react/hooks/useSendTransaction.en-US.mdx`
- [x] `react/hooks/useSignMessage.en-US.mdx`
- [x] `react/hooks/useSignTypedData.en-US.mdx`
- [x] `react/hooks/useSwitchNetwork.en-US.mdx`
- [x] `react/hooks/useToken.en-US.mdx`
- [x] `react/hooks/useTransaction.en-US.mdx`
- [x] `react/hooks/useWaitForTransaction.en-US.mdx`
- [x] `react/hooks/useWatchPendingTransactions.en-US.mdx`
- [x] `react/hooks/useWebSocketProvider.en-US.mdx`
- [x] `react/prepare-hooks/usePrepareContractWrite.en-US.mdx`
- [x] `react/prepare-hooks/usePrepareSendTransaction.en-US.mdx`
- [x] `react/providers/alchemy.en-US.mdx`
- [x] `react/providers/configuring-chains.en-US.mdx`
- [x] `react/providers/infura.en-US.mdx`
- [x] `react/providers/jsonRpc.en-US.mdx`
- [x] `react/providers/public.en-US.mdx`

### Migration Guide

- [ ] General
  - [x] Migrated from Ethers.js to viem.
  - [x] `BigNumber` has been removed in favor of browser-native `BigInt`.
    - [ ] Note that `BigInt` is not serializable (use `serialize`/`deserialize`).
  - [x] `gasLimit` is now `gas`.
  - [x] Ethers v5 "Provider" & "Signer" terminology is now known as viem's "Public Client" & "Wallet Client" terminology.
  - [x] Removed global type declaration for `window.ethereum`.
  - [x] Renamed `Ethereum` type to `WindowProvider`.
- [x] Actions/Hooks
  - [x] `connect`/`useConnect` @tmm
  - [x] `fetchBalance` / `useBalance` @tmm
  - [x] `fetchBlockNumber` / `useBlockNumber` @tmm
  - [x] `fetchEnsAvatar` / `useEnsAvatar` @tmm
  - [x] `fetchFeeData` / `useFeeData` @tmm
  - [x] `fetchSigner` / `useSigner` / `watchSigner` @tmm
  - [x] `fetchToken` / `useToken` @tmm
  - [x] `fetchTransaction` / `useTransaction` @tmm
  - [x] `useContract` / `getContract` @tmm
  - [x] `getProvider` / `useProvider` / `watchProvider` @tmm
  - [x] `getWebSocketProvider` / `useWebSocketProvider` / `watchWebSocketProvider` @tmm
  - [x] `multicall`
  - [x] `prepareSendTransaction` / `usePrepareSendTransaction`
  - [x] `prepareWriteContract` / `usePrepareContractWrite`
  - [x] `readContract` / `useContractRead`
  - [x] `readContracts` / `useContractReads` / `useContractInfiniteReads`
  - [x] `signMessage` / `useSignMessage`
  - [x] `signTypedData` / `useSignTypedData`
  - [x] `waitForTransaction`/`useWaitForTransaction`
  - [x] `watchPendingTransactions` / `useWatchPendingTransactions`
  - [x] `writeContract` / `useContractWrite`
  - [x] `watchContractEvent` / `useContractEvent`
- [x] `configureChains`
- [x] Connectors
  - [x] `getSigner` now returns a viem [`WalletClient`](https://viem.sh/docs/clients/wallet.html) instead of an Ethers `Signer`.
  - [x] Removed `Signer` generic.
- [x] Errors
  - [x] Removed `ChainDoesNotSupportMulticallError`, `ContractMethodDoesNotExistError`, `ContractMethodNoResultError`, `ContractMethodRevertedError`, `ContractResultDecodeError`, `ProviderRpcError`, `ResourceUnavailableError`

### Build

- [x] Resolve `pnpm build` issues

### CODE REVIEW

File: packages/core/src/actions/accounts/fetchBalance.test.ts [hunk 1]
```diff
@@ -16,16 +16,13 @@ describe('fetchBalance', () => {
...
-            "decimals": 18,
-            "formatted": "0.40742480512617271",
-            "symbol": "ETH",
-            "value": {
-              "hex": "0x05a776b39e3a7026",
-              "type": "BigNumber",
-            },
-          }
-        `)
+        {
```
Comment (jxom) [inline packages/core/src/actions/accounts/fetchBalance.test.ts:19]: Side-effect of updating Anvil block number.

**Comment by hammeiam** [comment]
Very excited for this, keep up the great work!

**Comment by hammeiam** [comment]
Watching all of these PR updates is better than Twitter.

**Comment by ennioVisco** [comment]
> Watching all of these PR updates is better than Twitter.

Hahaha I feel exactly the same! And seeing all these checks being added is sooo satisfying 😂

**Comment by jxom** [comment]
Closing in favor of #2235

**Pull request closed without merging**